docs/asyncio_imports.md

```diff
@@ -0,0 +1,38 @@
+---
+title: "Importing code with asyncio"
+---
+
+Determining when it is safe to import code when using asyncio can be tricky because two constraints need to be considered:
+
+- Importing code can do blocking I/O to load the files from the disk
+- Importing code in [cpython is not thread-safe](https://github.com/python/cpython/issues/83065)
+
+## Top-level imports
+
+Suppose your imports happen at the top-level (nearly all code at indentation level 0). Home Assistant will import your code before the event loop starts or import it in the import executor when your integration is loaded. In this case, you likely do not need to consider whether your imports are safe.
+
+## Imports outside of top-level
+
+If your imports are not happening at top-level, you must carefully consider each import, as the import machinery has to read the module from disk which does blocking I/O. If possible, it's usually best to change to a top-level import, as it avoids much complexity and the risk of mistakes. Importing modules is both CPU-intensive and involves blocking I/O, so it is crucial to ensure these operations are executed in the executor.
```

Comment:
I think there are two reasons for doing non module-level imports:
- Breaking circular import chains
  - Here, we could maybe mention imports which are only there for the benefit of the type check can be hidden behind an `if TYPE_CHECKING:` guard
  - We should probably also explain that if the import is imported also via an import chain starting at module-level of `__init__.py` it's safe to do a local import
- Avoiding importing a large set of modules which will mostly be unused
  - I think this is the case where it's relevant to manually do imports in the executor?